dev/PushNotifications/PushNotificationManager.cpp

```diff
@@ -0,0 +1,266 @@
+
+                PushNotificationTrigger trigger{};
+                builder.SetTrigger(trigger);
+
+                if (!IsPackagedProcess())
+                {
+                    throw winrt::hresult_not_implemented();
+                }
+
+                // In case the interface is not supported, let it throw.
+                auto builder5 = builder.as<winrt::IBackgroundTaskBuilder5>();
+                builder5.SetTaskEntryPointClsid(taskClsid);
+                winrt::com_array<winrt::IBackgroundCondition> conditions = details.GetConditions();
+                for (auto condition : conditions)
+                {
+                    builder.AddCondition(condition);
+                }
+            }
+        }
+
+        bool registeredWithCom = false;
+        bool registeredWithBackgroundTask = false;
+
+        auto scopeExitToCleanRegistrations = wil::scope_exit(
+            [&]()
+            {
+                if (registeredWithCom)
+                {
+                    LOG_IF_FAILED(::CoRevokeClassObject(cookie));
+                }
+
+                // Clean the task registration only if it was created during this call
+                if (registeredWithBackgroundTask)
+                {
+                    registeredTask.Unregister(true);
+                }
+
+            }
+        );
+
+        if (WI_IsFlagSet(registrationOption, PushNotificationRegistrationOption::ComActivator))
+        {
+            THROW_HR_IF_NULL(E_UNEXPECTED, g_waitHandleForArgs);
+
+            THROW_IF_FAILED(::CoRegisterClassObject(
+                taskClsid,
+                winrt::make<PushNotificationBackgroundTaskFactory>().get(),
+                CLSCTX_LOCAL_SERVER,
+                REGCLS_MULTIPLEUSE,
+                &cookie));
+
+            registeredWithCom = true;
+        }
+
+        if (builder)
+        {
+            auto registeredTaskFromBuilder = builder.Register();
+            registeredTask = registeredTaskFromBuilder.as<IBackgroundTaskRegistration>();
+            registeredWithBackgroundTask = true;
+        }
+
+        PushNotificationRegistrationToken token = { cookie, registeredTask };
+        scopeExitToCleanRegistrations.release();
+
+        return token;
+    }
+
+    void PushNotificationManager::UnregisterActivator(PushNotificationRegistrationToken const& token, PushNotificationRegistrationOption const& option)
+    {
+        THROW_HR_IF_NULL(E_INVALIDARG, token);
+        if (WI_IsFlagSet(option, PushNotificationRegistrationOption::PushTrigger))
+        {
+            auto taskRegistration = token.TaskRegistration();
+            THROW_HR_IF_NULL(HRESULT_FROM_WIN32(ERROR_NOT_FOUND), taskRegistration);
+            taskRegistration.Unregister(true);
+        }
+
+        if (WI_IsFlagSet(option, PushNotificationRegistrationOption::ComActivator) && token.Cookie())
+        {
+            LOG_IF_FAILED(::CoRevokeClassObject(static_cast<DWORD>(token.Cookie())));
```

Comment:
According to [the documentation for `CoRegisterClassObject`](https://docs.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coregisterclassobject), you shouldn't revoke the class object if there are outstanding locks on the class factory. We don't seem to check that. Is that okay?